### ISSUE TITLE
How are threshold signatures implemented?

### ISSUE THREAD

**Issue by Olshansk** [aptos-labs/aptos-core#499]
## ❓ Questions and Help

In [aptos-core/tree/main/consensus](https://github.com/aptos-labs/aptos-core/tree/main/consensus) it is implied that DKG is used to distributed keys that are later used for threshold signature aggregation: `This allows us to provide incentives to validators that contribute to quorum certificates. Aggregate signatures also do not require a complex [threshold key setup](https://www.cypherpunks.ca/~iang/pubs/DKG.pdf)[](https://github.com/aptos-labs/aptos-core/tree/main/consensus#implementation-details).`

However, the [QuorumCert](https://github.com/aptos-labs/aptos-core/blob/main/consensus/consensus-types/src/quorum_cert.rs) is using [LedgerInfoWithSignatures](https://github.com/aptos-labs/aptos-core/blob/main/types/src/ledger_info.rs), which simply uses a list of ed25519 signatures.

```rust
pub struct QuorumCert {
    /// The vote information certified by the quorum.
    vote_data: VoteData,
    /// The signed LedgerInfo of a committed block that carries the data about the certified block.
    signed_ledger_info: LedgerInfoWithSignatures,
}

---

impl LedgerInfoWithSignatures {
    pub fn new(
        ledger_info: LedgerInfo,
        signatures: BTreeMap<AccountAddress, Ed25519Signature>,
    ) -> Self {
        LedgerInfoWithSignatures::V0(LedgerInfoWithV0::new(ledger_info, signatures))
    }
...
```

In addition, I found the following requirement under [aptos-labs/aptos-core/tree/main/documentation/specifications/consensus](https://github.com/aptos-labs/aptos-core/tree/main/documentation/specifications/consensus): `ensure len(signed_ledger_info.signatures >= len(OnChainConfig.validators()) * 2 / 3 + 1`.

**Question**: Am I correct to understand that aggregation at the moment simply involves appending to a list and then looping through all the signatures?

**Comment by Olshansk** [comment]
I've also posted a link to this issue the #dev-discussion channel on aptos' discord per the recommendation in the docs.

**Comment by JoshLind** [comment]
'cc @zekun000 who is the expert here 😄 

My naive understanding is exactly how you describe it: each round leader will receive individual signatures from the participants (i.e., votes) in that round and those vote signatures are all simply grouped together in a list that can then be used to provide a quorum certificate for that round. But, I'll let @zekun000 confirm 😄

**Comment by zekun000** [comment]
we don't use threshold signatures for now, we have explored BLS in the past but not implemented in production. It's not standardized at the moment and the aggregation part is expensive from what I understand.

@alinush feel free to chime in about threshold signatures

**Comment by alinush** [comment]
@Olshansk, note that the document [you linked to](https://github.com/aptos-labs/aptos-core/tree/main/consensus) actually mentions _aggregate signatures_, not _threshold signatures_.

 > "Fourth, we use aggregate signatures that preserve the identity of validators who sign quorum certificates."

In an _aggregate signature scheme_, the verifier checks the signature against the _individual_ public keys of all signers involved in the aggregated signature. In other words, aggregate signature verification happens w.r.t. the public keys $(pk_{i_1}, pk_{i_2}, ..., pk_{i_k})$ of the involved signers $i_1, i_2, ..., i_k$.

In a _threshold signature scheme_, the verifier checks the signature against the public key $pk$ for the group of all signers, no matter which signers were involved in aggregating the signature.

I actually have updated some old slides to better explain this distinction. You can take a look [here](https://docs.google.com/presentation/d/1G4XGqrBLwqMyDQce_xpPQUEMOK4lFrneuvGYU3MVDsI/edit?usp=sharing). Hope it is useful.

To the best of my knowledge, the current aggregation in our "aggregate signature scheme' is very naive: just append all signatures together. This is because Ed25519 does not support efficient aggregation yet. (Although it does support some compression; see [this paper](https://eprint.iacr.org/2021/350))

**Comment by Olshansk** [comment]
Thanks for the details response @alinush.

Everything you said makes sense.

**Re aggregature signatures** - I think I interpreted "aggregate" as the term is commonly used with BLS to Schnorr signature aggregation, so there is only one signature to verify at the end.

**Re the use of Ed25519** - I saw in the docs that `Our goal was to choose a protocol that would initially support at least 100 validators and would be able to evolve over time to support 500–1,000 validators`, so should definitely work and scale well (time & size) in the initial phases.

I was wondering what the best place (this thread, discord, etc...) to potentially discuss other ideas down the line would be?
